File: golang/executors/gorm_ex/close_period.go
package gorm_ex

import (
	"bench-pg-go/executors/gorm_ex/model"
	"bench-pg-go/model/domain"
	"cloud.google.com/go/civil"
	"fmt"
	"github.com/shopspring/decimal"
	"gorm.io/gorm"
	"gorm.io/gorm/clause"
	"time"
)

type cdRecord struct {
	prevStock    decimal.Decimal `gorm:"column:prev_stock;not null"`
	receipt      decimal.Decimal `gorm:"column:receipt;not null"`
	consumption  decimal.Decimal `gorm:"column:consumption;not null"`
	diffAmount   decimal.Decimal `gorm:"column:diff_amount;not null"`
	nextStdPrice decimal.Decimal `gorm:"column:next_std_price;not null"`
}

func closePeriod(db *gorm.DB, period domain.Period, user domain.User) {

	userId := user.Id()

	materials := make([]model.MaterialEntity, 0)
	res := db.Select("id").Find(&materials)
	if res.Error != nil {
		panic(res.Error)
	}

	for _, material := range materials {
		materialId := material.Id
		currYp := period.YearPeriod()
		nextYp := period.NextPeriod().YearPeriod()
		prevYp := period.PrevPeriod().YearPeriod()
		updatedAt := time.Now()

		tx := db.Begin()
		if tx.Error != nil {
			panic(tx.Error)
		}

		var mp model.MaterialPeriodEntity
		res := tx.Select("std_price").
			Where("material_id = ? and period = ?", materialId, currYp).
			Clauses(clause.Locking{Strength: "UPDATE"}).
			First(&mp)
		if res.Error != nil {
			panic(res.Error)
		}

		var prevStock decimal.Decimal
		var receipt decimal.Decimal
		var consumption decimal.Decimal
		var diffAmount decimal.Decimal
		var nextStdPrice decimal.Decimal

		prevStockSubQuery := tx.Select("stock").
			Where("material_id = ? and period = ?", materialId, prevYp).
			Table("fin_material_periods").Scan(&prevStock)
		if prevStockSubQuery.Error != nil {
			panic(prevStockSubQuery.Error)
		}

		receiptSubQuery := tx.Select("COALESCE(sum(quantity),0)").
			Where("material_id = ? and period = ? and account_id = '10.01' and debt_credit = 'D'",
				materialId, currYp).
			Table("fin_ledger_items").Scan(&receipt)
		if receiptSubQuery.Error != nil {
			panic(receiptSubQuery.Error)
		}

		consumptionSubQuery := tx.Select("COALESCE(sum(quantity),0)").
			Where("material_id = ? and period = ? and account_id = '10.01' and debt_credit = 'C'",
				materialId, currYp).
			Table("fin_ledger_items").Scan(&consumption)
		if consumptionSubQuery.Error != nil {
			panic(consumptionSubQuery.Error)
		}

		diffAmountSubQuery := tx.Select("COALESCE(sum(amount),0)").
			Where("material_id = ? and period = ? and account_id = '10.01' and debt_credit = 'C'",
				materialId, currYp).
			Table("fin_ledger_items").Scan(&diffAmount)
		if diffAmountSubQuery.Error != nil {
			panic(diffAmountSubQuery.Error)
		}

		nextStdPriceSubQuery := tx.Select("std_price").
			Where("material_id = ? and period = ?", materialId, nextYp).
			Table("fin_material_periods").Scan(&nextStdPrice)
		if nextStdPriceSubQuery.Error != nil {
			panic(nextStdPriceSubQuery.Error)
		}

		//var cd cdRecord
		//res = tx.Select("(?) as prev_stock, (?) as receipt, (?) as consumption, (?) as diff_amount, (?) as next_std_price",
		//	prevStockSubQuery, receiptSubQuery, consumptionSubQuery, diffAmountSubQuery, nextStdPriceSubQuery).
		//	Table("fin_material_periods").
		//	Scan(&cd)
		//if res.Error != nil {
		//	panic(res.Error)
		//}

		stdPrice := mp.StdPrice
		//prevStock := cd.prevStock
		//receipt := cd.receipt
		//consumption := cd.consumption
		//diffAmount := cd.diffAmount
		//nextStdPrice := cd.nextStdPrice

		// Общая стоимость: ta2 = s1 * sp2 + r2 * sp2 +d2
		totalAmount := prevStock.Mul(stdPrice).Add(receipt.Mul(stdPrice)).Add(diffAmount)
		// Общий запас: ts2 = s1 + r2
		totalStock := prevStock.Add(receipt)
		// Факт цена: ap = ta2 / ts2
		actualPrice := totalAmount.Div(totalStock)
		// Запас на конец периода: s3 = s1 + r2 + c2
		currStock := prevStock.Add(receipt).Add(consumption)

		docNo := fmt.Sprintf("CLOSE-%s-%s", period.String(), materialId)

		// Отклонения на запас: ds2 = d2 * s3 / (s1 + r2)
		// Проводка Dt 10.01 Ct 10.02 ds2
		diffToStockAmount := diffAmount.Mul(currStock).Div(prevStock.Add(receipt))

		postDifferences(
			tx,
			period.LastDate(),
			docNo,
			materialId,
			diffToStockAmount,
			domain.INVENTORY_ACCOUNT,
			domain.INVENTORY_DIFF_ACCOUNT,
			userId,
			updatedAt,
		)

		// Отклонения на COGS: dc2 = d2 - ds2
		// Проводка Dt 90.02 Ct 10.02 dc2
		diffToCogsAmount := diffAmount.Sub(diffToStockAmount)
		postDifferences(
			tx,
			period.LastDate(),
			docNo,
			materialId,
			diffToCogsAmount,
			domain.COGS_ACCOUNT,
			domain.INVENTORY_DIFF_ACCOUNT,
			userId,
			updatedAt,
		)

		docNo = fmt.Sprintf("OPEN-%s-%s", period.String(), materialId)

		// Отклонение след. периода ds3 = s3 * sp2 + ds2 - s3 * sp3
		// Проводка Dt 10.02 Ct 10.01 ds3
		nextDiffAmount := currStock.Mul(stdPrice).Add(diffToStockAmount).Sub(currStock.Mul(nextStdPrice))
		postDifferences(
			tx,
			period.NextPeriod().FirstDate(),
			docNo,
			materialId,
			nextDiffAmount,
			domain.INVENTORY_DIFF_ACCOUNT,
			domain.INVENTORY_ACCOUNT,
			userId,
			updatedAt,
		)

		// update next_std_price material period
		res = tx.Model(&model.MaterialEntity{}).
			Where("id = ?", materialId).
			Updates(map[string]interface{}{
				"next_std_price": nextStdPrice,
				"updated_by":     userId,
				"updated_at":     updatedAt,
			})
		if res.Error != nil {
			panic(res.Error)
		}

		// update actual price
		res = tx.Model(&model.MaterialPeriodEntity{}).
			Where("material_id = ? and period = ?", materialId, currYp).
			Updates(map[string]interface{}{
				"actual_price": actualPrice,
				"updated_by":   userId,
				"updated_at":   updatedAt,
			})
		if res.Error != nil {
			panic(res.Error)
		}

		res = tx.Commit()
		if res.Error != nil {
			panic(res.Error)
		}
	}
}

func postDifferences(
	tx *gorm.DB,
	postingDate civil.Date,
	docNo string,
	materialId string,
	amount decimal.Decimal,
	debtAccount string,
	creditAccount string,
	userId string,
	updatedAt time.Time,
) {

	if amount.Equal(decimal.Zero) {
		return
	}

	if amount.LessThan(decimal.Zero) {
		amount = amount.Neg()
		acc := debtAccount
		debtAccount = creditAccount
		creditAccount = acc
	}

	period := domain.PeriodFromDate(postingDate)

	lines := []model.LedgerItemEntity{
		{
			Id:                0,
			Period:            period.YearPeriod(),
			DocNo:             docNo,
			PostingDate:       postingDate.String(),
			AccountId:         debtAccount,
			MaterialId:        &materialId,
			BusinessPartnerId: nil,
			DebtCredit:        domain.DEBT,
			Amount:            amount,
			Debt:              amount,
			Credit:            decimal.Zero,
			Quantity:          decimal.Zero,
			UpdatedBy:         userId,
			UpdatedAt:         updatedAt,
		},
		{
			Id:                0,
			Period:            period.YearPeriod(),
			DocNo:             docNo,
			PostingDate:       postingDate.String(),
			AccountId:         creditAccount,
			MaterialId:        &materialId,
			BusinessPartnerId: nil,
			DebtCredit:        domain.CREDIT,
			Amount:            amount.Neg(),
			Credit:            decimal.Zero,
			Debt:              amount,
			Quantity:          decimal.Zero,
			UpdatedBy:         userId,
			UpdatedAt:         updatedAt,
		},
	}

	res := tx.Create(&lines)
	if res.Error != nil {
		panic(res.Error)
	}

}

File: rust/src/executors/diesel/close_period.rs
use crate::model::domain::{COGS_ACCOUNT, CREDIT, DEBT, INVENTORY_ACCOUNT, INVENTORY_DIFF_ACCOUNT, Period, User,};
use chrono::{NaiveDate, Utc};
use diesel::dsl::*;
use diesel::prelude::*;
use diesel::r2d2::{ConnectionManager, PooledConnection};
use diesel::result::Error;
use rust_decimal::Decimal;
use diesel::{PgConnection, RunQueryDsl};
use crate::executors::diesel::schema::{fin_ledger_items, fin_material_periods, fin_materials};

pub async fn close_period(
    conn: &mut PooledConnection<ConnectionManager<PgConnection>>,
    period: Period,
    user: User,
) {
    let user_id = user.id();

    let materials = fin_materials::dsl::fin_materials
        .select(fin_materials::dsl::id)
        .load::<String>(conn)
        .expect("Error loading materials");

    for material_id in materials {
        let curr_yp = period.year_period();
        let next_yp = period.next_period().year_period();
        let prev_yp = period.prev_period().year_period();

        conn.transaction(|tx| {

            let std_price = fin_material_periods::dsl::fin_material_periods
                .select(
                    fin_material_periods::dsl::std_price,
                )
                .filter(fin_material_periods::dsl::material_id.eq(&material_id))
                .filter(fin_material_periods::dsl::period.eq(curr_yp))
                .for_update()
                .first::<Decimal>(tx)
                .expect("Error getting material period");

            let prev_stock_res = fin_material_periods::dsl::fin_material_periods
                .select(
                    fin_material_periods::dsl::stock,
                )
                .filter(fin_material_periods::dsl::material_id.eq(&material_id))
                .filter(fin_material_periods::dsl::period.eq(prev_yp))
                .first::<Decimal>(tx);

            let prev_stock = match prev_stock_res {
                Ok(ps) => { ps}
                Err(Error::NotFound) => {Decimal::from(0)}
                _ => panic!("Error calculate closing data")
            };

            let receipt = fin_ledger_items::dsl::fin_ledger_items
                .select(
                    sum(fin_ledger_items::dsl::quantity),
                )
                .filter(fin_ledger_items::dsl::period.eq(curr_yp))
                .filter(fin_ledger_items::dsl::material_id.eq(&material_id))
                .filter(fin_ledger_items::dsl::account_id.eq("10.01"))
                .filter(fin_ledger_items::dsl::debt_credit.eq("D"))
                .first::<Option<Decimal>>(tx)
                .expect("Error calculate closing data")
                .unwrap_or(Decimal::from(0));

            let consumption = fin_ledger_items::dsl::fin_ledger_items
                .select(
                    sum(fin_ledger_items::dsl::quantity),
                )
                .filter(fin_ledger_items::dsl::period.eq(curr_yp))
                .filter(fin_ledger_items::dsl::material_id.eq(&material_id))
                .filter(fin_ledger_items::dsl::account_id.eq("10.01"))
                .filter(fin_ledger_items::dsl::debt_credit.eq("C"))
                .first::<Option<Decimal>>(tx)
                .expect("Error calculate closing data")
                .unwrap_or(Decimal::from(0));

            let diff_amount = fin_ledger_items::dsl::fin_ledger_items
                .select(
                    sum(fin_ledger_items::dsl::amount),
                )
                .filter(fin_ledger_items::dsl::period.eq(curr_yp))
                .filter(fin_ledger_items::dsl::material_id.eq(&material_id))
                .filter(fin_ledger_items::dsl::account_id.eq("10.02"))
                .first::<Option<Decimal>>(tx)
                .expect("Error calculate closing data")
                .unwrap_or(Decimal::from(0));

            let next_std_price = fin_material_periods::dsl::fin_material_periods
                .select(
                    fin_material_periods::dsl::std_price,
                )
                .filter(fin_material_periods::dsl::material_id.eq(&material_id))
                .filter(fin_material_periods::dsl::period.eq(next_yp))
                .first::<Decimal>(tx)
                .expect("Error calculate closing data");

            // Общая стоимость: ta2 = s1 * sp2 + r2 * sp2 +d2
            let total_amount = prev_stock * std_price + receipt * std_price + diff_amount;
            // Общий запас: ts2 = s1 + r2
            let total_stock = prev_stock + receipt;
            // Факт цена: ap = ta2 / ts2
            let actual_price = total_amount / total_stock;
            // Запас на конец периода: s3 = s1 + r2 + c2
            let curr_stock = prev_stock + receipt + consumption;

            let doc_no = format!("CLOSE-{}-{}", period, material_id);

            // Отклонения на запас: ds2 = d2 * s3 / (s1 + r2)
            // Проводка Dt 10.01 Ct 10.02 ds2
            let diff_to_stock_amount = diff_amount * curr_stock / (prev_stock + receipt);
            post_differences(
                tx,
                period.last_date(),
                &doc_no,
                &material_id,
                &diff_to_stock_amount,
                INVENTORY_ACCOUNT,
                INVENTORY_DIFF_ACCOUNT,
                &user_id,
            );

            // Отклонения на COGS: dc2 = d2 - ds2
            // Проводка Dt 90.02 Ct 10.02 dc2
            let diff_to_cogs_amount = diff_amount - diff_to_stock_amount;
            post_differences(
                tx,
                period.last_date(),
                &doc_no,
                &material_id,
                &diff_to_cogs_amount,
                COGS_ACCOUNT,
                INVENTORY_DIFF_ACCOUNT,
                &user_id,
            );

            let doc_no = format!("OPEN-{}-{}", next_yp, &material_id);
            // Отклонение след. периода ds3 = s3 * sp2 + ds2 - s3 * sp3
            // Проводка Dt 10.02 Ct 10.01 ds3
            let next_diff_amount =
                curr_stock * std_price + diff_to_stock_amount - curr_stock * next_std_price;
            post_differences(
                tx,
                period.next_period().first_date(),
                &doc_no,
                &material_id,
                &next_diff_amount,
                INVENTORY_DIFF_ACCOUNT,
                INVENTORY_ACCOUNT,
                &user_id,
            );

            // update next_std_price material period
            diesel::update(
                fin_materials::dsl::fin_materials
                    .find(&material_id),
            )
                .set((
                    fin_materials::dsl::next_std_price.eq(actual_price),
                    fin_materials::dsl::updated_by.eq(&user_id),
                    fin_materials::dsl::updated_at.eq(&Utc::now()),
                ))
                .execute(tx)
                .expect("Error updating material period");

            // update actual price
            diesel::update(
                fin_material_periods::dsl::fin_material_periods
                    .find((&material_id, period.year_period())),
            )
                .set((
                    fin_material_periods::dsl::actual_price.eq(actual_price),
                    fin_material_periods::dsl::updated_by.eq(&user_id),
                    fin_material_periods::dsl::updated_at.eq(&Utc::now()),
                ))
                .execute(tx)
                .expect("Error updating material period");

            Ok::<(), Error>(())
        })
            .expect(
            format!(
                "Cannot commit close period transaction for material {} in period {}",
                &material_id, period,
            )
                .as_str(),
        );
    }
}

fn post_differences(
    tx: &mut PooledConnection<ConnectionManager<PgConnection>>,
    posting_date: NaiveDate,
    doc_no: &String,
    material_id: &String,
    amount: &Decimal,
    debt_account: &str,
    credit_account: &str,
    user_id: &String,
) {
    if *amount == Decimal::from(0) {
        return;
    }

    let (amount, debt_account, credit_account) = if *amount < Decimal::from(0) {
        (&-amount, credit_account, debt_account)
    } else {
        (amount, debt_account, credit_account)
    };
    let period = Period::from(posting_date);
    let updated_at = Utc::now();

    diesel::insert_into(fin_ledger_items::dsl::fin_ledger_items)
        .values((
            fin_ledger_items::dsl::period.eq(period.year_period()),
            fin_ledger_items::dsl::doc_no.eq(doc_no),
            fin_ledger_items::dsl::posting_date.eq(&posting_date),
            fin_ledger_items::dsl::account_id.eq(debt_account),
            fin_ledger_items::dsl::material_id.eq(material_id),
            fin_ledger_items::dsl::debt_credit.eq(DEBT),
            fin_ledger_items::dsl::amount.eq(&amount),
            fin_ledger_items::dsl::debt.eq(&amount),
            fin_ledger_items::dsl::credit.eq(Decimal::from(0),),
            fin_ledger_items::dsl::quantity.eq(Decimal::from(0)),
            fin_ledger_items::dsl::updated_by.eq(&user_id),
            fin_ledger_items::dsl::updated_at.eq(&updated_at),
        ))
        .execute(tx)
        .expect("Cannot insert post_differences line 1");

    diesel::insert_into(fin_ledger_items::dsl::fin_ledger_items)
        .values((
            fin_ledger_items::dsl::period.eq(period.year_period()),
            fin_ledger_items::dsl::doc_no.eq(doc_no),
            fin_ledger_items::dsl::posting_date.eq(&posting_date),
            fin_ledger_items::dsl::account_id.eq(credit_account),
            fin_ledger_items::dsl::material_id.eq(material_id),
            fin_ledger_items::dsl::debt_credit.eq(CREDIT),
            fin_ledger_items::dsl::amount.eq(&-amount),
            fin_ledger_items::dsl::debt.eq(Decimal::from(0)),
            fin_ledger_items::dsl::credit.eq(&amount),
            fin_ledger_items::dsl::quantity.eq(Decimal::from(0)),
            fin_ledger_items::dsl::updated_by.eq(&user_id),
            fin_ledger_items::dsl::updated_at.eq(&updated_at),
        ))
        .execute(tx)
        .expect("Cannot insert post_differences line 2");
}
